### PR TITLE
initial API for creating and retrieving bookmarks

### PR DESCRIPTION
**Pull request by clemux**
Status: can be integrated, but not feature-complete yet.

### CODE REVIEW

File: archivy/models.py [hunk 1]
```diff
@@ -79,17 +68,16 @@ def process_bookmark_url(self):
         try:
             self.content = self.extract_content(parsed_html)
         except Exception:
-            flash(f"Could not extract content from {self.url}\n")
-            return
+            raise RuntimeError(f"Could not extract content from {self.url}\n")
```
Comment (Uzay-G) [inline archivy/models.py:71]: Makes more sense to keep the flash here so the user can see it, it's not a backend error it's the input that has a problem.

File: archivy/models.py [hunk 2]
```diff
@@ -1,23 +1,57 @@
...
 import frontmatter
-from flask import flash
+import html2text
+import requests
+import validators
+from attr import attrs, attrib
+from attr.validators import instance_of, optional
 from bs4 import BeautifulSoup
 from flask import current_app
+from flask import flash
```
Comment (Uzay-G) [inline archivy/models.py:13]: Could we combine the two flask imports here?

File: archivy/models.py [hunk 3]
```diff
@@ -1,23 +1,57 @@
...
 
 from archivy import extensions
-from archivy.search import add_to_index
 from archivy.data import create
+from archivy.search import add_to_index
 
 
+# TODO: use this as 'type' field
```
Comment (Uzay-G) [inline archivy/models.py:20]: This Enum is a good idea.

File: archivy/models.py [hunk 4]
```diff
@@ -1,23 +1,57 @@
...
+#     BOOKMARK = 'bookmark'
+#     POCKET_BOOKMARK = 'bookmark imported from pocket'
+#     NOTE = 'note'
+#     PROCESSED_DATAOBJ = 'bookmark that has been processed'
+
+@attrs(kw_only=True)
 class DataObj:
     __searchable__ = ["title", "content", "desc", "tags"]
 
+    id: Optional[int] = attrib(validator=optional(instance_of(int)),
```
Comment (Uzay-G) [inline archivy/models.py:31]: I like this clear declaration of the attrs as it provides a clearer structure.

Comment (Uzay-G) [reply archivy/models.py:31]: Does this totally eliminate the need for a `init` function and instead simply maps the keyword arguments if they exist?

Comment (clemux) [reply archivy/models.py:31]: Yes, attts creates the constructor. With `kw_only=True`, all parameters of the constructor are keyword arguments.

File: archivy/models.py [hunk 5]
```diff
@@ -66,34 +99,6 @@ def extract_content(self, beautsoup):
 
         return html2text.html2text(str(beautsoup))
 
-    def __init__(self, **kwargs):
-
-        # data has already been processed
-        if kwargs["type"] == "processed-dataobj":
```
Comment (Uzay-G) [inline archivy/models.py:72]: How will this work with the class method that is called if we are using an already existing dataobj?

**Comment by Uzay-G** [comment]
I'd like to get the model changes merged. Could you maybe make a new pr that takes the changes here without the api and incorporates them?

**Comment by clemux** [comment]
Sure, in an hour or two!

**Comment by clemux** [comment]
Don't hesitate to cherry pick from my branches, though ;)

**Comment by Uzay-G** [comment]
I noticed you rebased the branch. If you think it's ready, please don't hesitate to ping me and I'll get to work reviewing the pr :smile;

**Comment by Uzay-G** [comment]
@clemux how are you doing with this? I'd be happy to help with any problems / questions you might have. Especially as this will help with allowing the user to code around with the platform

**Comment by clemux** [comment]
You can start reviewing the PR. There's a FIXME comment in the `GET /api/bookmarks` route, and `PUT /api/bookmarks/<id>` is not implemented, though.

**Comment by Uzay-G** [comment]
Is there a reason you made this api only for `bookmarks`? We could generalize it to dataobjs in general and provide a `type` param to only gather certain results.

Here is what I can implement:

- The FIXME
- the new api route
- generalizing bookmarks to dataobjs

If you've already started working on one of the first two, please tell me.

Is this good with you?

**Comment by clemux** [comment]
I wanted to limit the size of the changes so that we can integrate them into  the main branch as soon as possible, because parallel feature branches are a pain to maintain. 

I probably won't have time to work on this until Saturday, so feel free to add whatever you want :)

**Comment by Uzay-G** [comment]
Is it ok if I push commits onto your branch? I won't force-push, just to implement fixes and changes.

Is there any better way I should do things?

**Comment by clemux** [comment]
Yes of course, I don't know how other projects do that, the only alternative I can see is that you open a PR in repository, and I'm not sure that'll be helpful.

**Comment by Uzay-G** [comment]
Sorry I couldn't get to this yet - plan on doing it this weekend :+1:

**Comment by Uzay-G** [comment]
Very excited about getting this merged!

**Pull request merged**